### PR TITLE
Switches to using DateTimeOffset for all dates

### PR DESCRIPTION
**Pull request by mscottford**
Extracts `DateTimeOffset` related changes from #243. This should address the issues reported in #256 and #257.

The changes have been simplified into a single commit with the following commit message:

* The `DateTimeOffset` type is used for all internal date representations in the codebase.
* This updates repositories to read to date time values that include the timezone offset. In the case of the RubyRepository, this required switching to a different method of retrieving the version information. If no timezone is available, then we assume that the values are in UTC (no timezone offset).
* This updates the way that date ranges are computed using a source code repository's change history. Now the logic uses the timezone of the requestor (via the `asOf` value) as the timezone that should be used for all dates in the supplied range. And to avoid an edge case, the date of the first commit in the change history is always the first date in the range. The next date is the start of the month that follows that date, and then the dates increment one month at a time until the `asOf` date time is reached.
* This pull request also addresses many issues that were reported by CodeClimate in the test suite. There is still more work to be done to address the issues in test land, but this pull request brings things a lot closer.

### CODE REVIEW

**Comment by mscottford** [comment]
This is ready for review.

**Comment by mscottford** [comment]
@mrbiggred Are you able to review this or should I ask someone else to take a peek?

**Comment by mrbiggred** [comment]
@mscottford thanks for reminding me.  I've started on the review.

File: Corgibytes.Freshli.Lib.Test/Unit/Util/DateTimeExtensionsTest.cs
Comment (mrbiggred) [inline Corgibytes.Freshli.Lib.Test/Unit/Util/DateTimeExtensionsTest.cs:224]: Can you use the `DateBuilder.cs` Class?

File: Corgibytes.Freshli.Lib.Test/Unit/Util/DateTimeExtensionsTest.cs
Comment (mrbiggred) [inline Corgibytes.Freshli.Lib.Test/Unit/Util/DateTimeExtensionsTest.cs:238]: Same thoughts for these DateTimeOffset extension tests as the Repository tests.  What do you think about having tests for the each of the extensions that call common test methods?

File: Corgibytes.Freshli.Lib/Util/DateTimeExtensions.cs
Comment (mrbiggred) [inline Corgibytes.Freshli.Lib/Util/DateTimeExtensions.cs:4]: Should this be changed to `DateTimeOffsetExtensions`?

**Review by mrbiggred — CHANGES_REQUESTED** [review]
No issues with the changes from DateTime to DateTimeOffset.  I do have some coding style questions but we can discuss that in Slack and aren't really important to this PR.  There are also some minor possible dead code, methods not used when they should, and other minor things.  See inline comments for details.

My main feedback is too think about de-normalizing some of the tests.  I think some of the tests are too refactored/normalized.  They could be easier to understand if there where de-normalized a bit.  See inline comments for more details.

As always let me know what you think and thanks for fixing this annoying date/time zone issue.  Glad it was found no while we are still in alpha.

File: Corgibytes.Freshli.Lib/Util/DateTimeExtensions.cs
Comment (mscottford) [reply Corgibytes.Freshli.Lib/Util/DateTimeExtensions.cs:4]: Good point. I'll fix that.

File: Corgibytes.Freshli.Lib.Test/Integration/PackageRepositoryTest.cs
Comment (mscottford) [reply Corgibytes.Freshli.Lib.Test/Integration/PackageRepositoryTest.cs:240]: I've been thinking about this quite a bit, and I've done a lot of research to explore many different options. I think what I'm going to do is create a `RespositoryTestFixture<>` abstract base class that contains the duplicated test methods. The different implementations for `RubyGemsRepository` and the like will end up looking something like this:

```
namespace Corgibytes.Freshli.Lib.Test.Integration.Languages.Ruby
{
    public class RubyGemsRepositoryTest: RepositoryTestFixture<RubyGemsRepository>
    {
        protected override IRepository Repository => new RubyGemsRepository();

        protected override TheoryData<IList<string>, IList<int>, string> DataForTestingLatest => new()
        {
                {
                     new[] {"tzinfo", "1.2.7"},
                     new[] {2020, 04, 02, 21, 42, 11},
                     "1.2.7"
                },
                {
                      new[] {"git", "1.6.0.pre1"},
                      new[] {2020, 01, 20, 20, 50, 43},
                      "1.6.0.pre1"
                },
                {
                      new[] {"monolog/monolog", "1.11.0"},
                      new[] {2014, 09, 30, 13, 30, 58},
                      "1.11.0"
                }
        };
        
        // other methods redacted but they would also provide test data
}
```

What do you think about that? Would you consider that a better choice than what's been implemented in this pull request?

File: Corgibytes.Freshli.Lib.Test/Integration/AssertIPackageRepository.cs
Comment (mscottford) [reply Corgibytes.Freshli.Lib.Test/Integration/AssertIPackageRepository.cs:5]: Yep. Looks like that can be safely removed.

File: Corgibytes.Freshli.Lib.Test/Unit/Util/DateTimeExtensionsTest.cs
Comment (mscottford) [reply Corgibytes.Freshli.Lib.Test/Unit/Util/DateTimeExtensionsTest.cs:238]: I think that something similar to my suggestion for the IRepository testers could be used here as well.

```csharp
public class TestDateTimeOffsetExtensions
{
    public class ToEndOfDay: DateModificationExtensionMethodTestFixture
    {
        protected override DateTimeOffset InvokeMethod(DateTimeOffset value) => value.ToEndOfDay();

        protected override TheoryData<IList<int>, IList<int>, int> DataForTestingExtensionMethod => new()
        {
            { 
                 new[] {2020, 1, 5, 23, 59, 59, 999, 9998},
                 new[] {2020, 1, 6, 23, 59, 59, 999, 9999},
                 -1
            },
            // additional test data snipped
        }
    }

    public class ToStartOfDay: DateModificationExtensionMethodTestFixture
    {
        protected override DateTimeOffset InvokeMethod(DateTimeOffset value) => value.ToStartOfDay();

        // contents snipped
    }

}

Would that work?
```

File: Corgibytes.Freshli.Lib.Test/Integration/PackageRepositoryTest.cs
Comment (mrbiggred) [reply Corgibytes.Freshli.Lib.Test/Integration/PackageRepositoryTest.cs:240]: I like that idea.

File: Corgibytes.Freshli.Lib.Test/Unit/Util/DateTimeExtensionsTest.cs
Comment (mrbiggred) [reply Corgibytes.Freshli.Lib.Test/Unit/Util/DateTimeExtensionsTest.cs:238]: I like that idea.  That way everyone can easily find the test file.  They can also add tests specific to that extension if needed.

**Comment by mscottford** [comment]
I'm working on getting master merged into this branch so that it runs without test failures, and then I'm going to tackle the code review comments.

**Comment by mscottford** [comment]
Yay! All of the tests are passing. Time to fix that one pesky Code Climate issue and then on to responding to the code review comments.

**Comment by mrbiggred** [comment]
@mscottford glad you got the tests passing and master merged into your branch.  It also looks you figured out the Approval Tests issue.  I'm posting a previous Approval Test fix for dealing with different operating system line endings for posterity.

https://github.com/corgibytes/freshli-lib/pull/327#issuecomment-893836192

https://github.com/corgibytes/freshli-lib/pull/327/commits/a779b753699225365d4af004a4d2b084212965ea

**Comment by mscottford** [comment]
That change wasn't removed. Just expanded upon.

**Comment by mscottford** [comment]
Because of the build matrix changes that are included in this branch, we'll probably have to override the merge block, and then update settings that control the list of things that need to pass before a PR is allowed to be merged. I'm guessing that "build" doesn't automatically expand out to a matrix of the same name.

**Comment by mscottford** [comment]
I added a few more jobs to the pipeline and I named one of them `build`. That should fix the expected status, I think.

File: Corgibytes.Freshli.Lib.Test/Integration/PackageRepositoryTest.cs
Comment (mscottford) [reply Corgibytes.Freshli.Lib.Test/Integration/PackageRepositoryTest.cs]: I’ve completed this part of the review feedback.

File: Corgibytes.Freshli.Lib.Test/Integration/Languages/Ruby/RubyGemsRepositoryProperlyHandlesRateLimitsTest.cs
Comment (mscottford) [inline Corgibytes.Freshli.Lib.Test/Integration/Languages/Ruby/RubyGemsRepositoryProperlyHandlesRateLimitsTest.cs:89]: Some odd formatting here.

Comment (mscottford) [reply Corgibytes.Freshli.Lib.Test/Integration/Languages/Ruby/RubyGemsRepositoryProperlyHandlesRateLimitsTest.cs:89]: I would have expected the auto format job to correct this.

File: Corgibytes.Freshli.Lib.Test/Integration/Languages/Ruby/RubyGemsRepositoryTest.cs
Comment (mscottford) [inline Corgibytes.Freshli.Lib.Test/Integration/Languages/Ruby/RubyGemsRepositoryTest.cs:45]: I kind of like that the warning messages show up here. Gives me an extra incentive to address them.

File: Corgibytes.Freshli.Lib.Test/Integration/Languages/Ruby/RubyGemsRepositoryTest.cs
Comment (mscottford) [inline Corgibytes.Freshli.Lib.Test/Integration/Languages/Ruby/RubyGemsRepositoryTest.cs:48]: Wrong word. Should be `intentionally` not `initially`.

File: Corgibytes.Freshli.Lib.Test/Integration/LibYearCalculatorTest.cs
Comment (mscottford) [reply Corgibytes.Freshli.Lib.Test/Integration/LibYearCalculatorTest.cs:27]: Good call.

File: Corgibytes.Freshli.Lib.Test/Integration/init-coverage.txt
Comment (mscottford) [inline Corgibytes.Freshli.Lib.Test/Integration/init-coverage.txt:16]: Reminder to myself: delete this file before merge.

File: Corgibytes.Freshli.Lib/Languages/CSharp/NuGetRepository.cs
Comment (mscottford) [inline Corgibytes.Freshli.Lib/Languages/CSharp/NuGetRepository.cs:64]: I need to add a package repository test for this class.

**Comment by mscottford** [comment]
Looks Linda the addition of the job named `build` has done the trick. 

There is still more work to do on this PR, and it’s a whole lot closer than it was before.

**Comment by mscottford** [comment]
Okay, @mrbiggred. This is ready for another review. I _think_ I’m done. Yay!

**Comment by mscottford** [comment]
Also, I really appreciate you putting so much thought into your reviews. I will admit that I was initially frustrated when I read your review comments.

The purpose of `Corgibytes.Xunit.Extensions` is to allow using an object instance member as a source for the data that’s passed into a method that’s decorated with `[Theory]`. Xunit’s out of the box behavior only permits using a static class member for specifying those sources.

Also, a goal that I was reaching for (with both this implementation and the previous one), is a way for a class that implements IRepository to validate that the implementation meets the expectations of the rest of the library. I was trying to create a centralized test fixture that would validate all of the methods on the interface, using data that’s specifically for the instance of `IRepository` that is under test.

I considered adopting the approach of creating an assertion library to achieve that, but I decided that I preferred the route of having one class that “knows” how to test an instance of `IRepository` over just having assertion methods to support validating.

I also started to find specifying test data in the form of method attributes rather constraining. All of the values that are passed into an attribute’s constructor when it is invoked as `[Thing(values)]` have to be constants or you get a compiler error. So constructing an instance of `DateTimeOffset` is not allowed. (See explanation here: https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/language-specification/attributes — Search for “An expression E is an attribute_argument_expression if all of the following statements are true:”) (You can also search StackOverflow for questions about people trying to do this.) I was not a huge fan of the work arounds that I created to get around the issue (namely, pass in the arguments to the `DateTimeOffset` constructor as an array, because arrays are allowed). This made using the MemberData approach that Xunit applies appealing, but the requirement to have that member be a static one generated another problem.

An instance method in an abstract class has no way of polymorphically accessing static members that are defined in child classes. This meant that the abstract base version of the test class was not allowed to delegate the generation of the `MemberData` to one of it’s concrete descendants. There is a proposal to the C# language to allow the specification of static members as part of an interface, but it is still under active discussion.

I ran into several discussions when searching for a solution, which lead me to believe that I was not the only person out there who was trying to adopt this testing pattern. (I kind of expected this particular pattern to be mentioned in the book “xUnit Test Patterns”, but it appears to have been written when MbUnit was one of the few libraries that supported specifying parameterized tests in this way. https://learning.oreilly.com/library/view/xunit-test-patterns/9780131495050/) This got me thinking that it would be fun to share my solution to the problem in the form of another library. There is a vast ecosystem of Xunit extension projects.

So that’s the rationale for how the code is structured the way it is. I hope that helps. With all of that in mind, do you still object to the pull request in it’s current state?

P.S. I originally posted this just in Slack, and then I realized that it made more sense to capture this rationale here.

**Comment by mscottford** [comment]
And regarding the logging, the parts of Xunit that are being extended do not provide a mechanism for logging, so that's why I created the solution that you see in the extension. It's purely for debugging purposes, and is off by default.

**Comment by mscottford** [comment]
I’m going to work on breaking this up with the plan that the commits related to addressing the CodeClimate findings in the tests suite are going to be abandoned. Hopefully, that will allow us to move forward with these changes in a more incremental way. Once that’s all done, then I’ll close this pull request and create companion “tech-debt” or “cleanup” tickets to address the issues that have been reported by CodeClimate.

**Comment by mscottford** [comment]
I got `main` merged in and all of the unit and integration tests are passing. I’ll tackle the rest later.

**Comment by mscottford** [comment]
Per a verbal okay from @mrbiggred, I’m going to merge this into main once the tests are passing again.

**Pull request merged**
